python_modules/dagster/dagster/_core/execution/context/compute.py

```diff
@@ -1301,10 +1302,113 @@ def set_requires_typed_event_stream(self, *, error_message: Optional[str] = None
+            ("assets_def", PublicAttr[Optional[AssetsDefinition]]),
+            ("code_version", PublicAttr[Optional[str]]),
+            ("provenance", PublicAttr[Optional[DataProvenance]]),
+        ],
+    )
+):
+    def __new__(
+        cls,
+        key: AssetKey,
+        code_version: Optional[str] = None,
+        provenance: Optional[DataProvenance] = None,
+        assets_def: Optional[AssetsDefinition] = None,
+    ):
+        return super(AssetInfo, cls).__new__(
+            cls, key=key, code_version=code_version, provenance=provenance, assets_def=assets_def
+        )
+
+
+class RunInfo(
+    NamedTuple(
+        "_RunInfo",
+        [
+            ("run_id", PublicAttr[str]),
+            ("dagster_run", PublicAttr[DagsterRun]),
+            ("run_config", PublicAttr[Mapping[str, object]]),
+            ("retry_number", PublicAttr[int]),
+        ],
+    )
+):
+    def __new__(
+        cls,
+        run_id: str,
+        dagster_run: DagsterRun,
+        run_config: Mapping[str, object],
+        retry_number: int,
+    ):
+        return super(RunInfo, cls).__new__(
+            cls,
+            run_id=run_id,
+            dagster_run=dagster_run,
+            run_config=run_config,
+            retry_number=retry_number,
+        )
+
+
 class AssetExecutionContext(OpExecutionContext):
     def __init__(self, step_execution_context: StepExecutionContext):
         super().__init__(step_execution_context=step_execution_context)
 
+        self._run_info = RunInfo(
+            run_id=self.run_id,
+            dagster_run=self.run,
+            run_config=self.run_config,
+            retry_number=self.retry_number,
+        )
+
+        self._asset_info = None
+        self._asset_info_by_key = None
+
+    @property
+    def asset_info(self) -> AssetInfo:
+        if len(self.assets_def.keys_by_output_name.keys()) > 1:
+            raise DagsterInvariantViolationError(
+                "Cannot call `context.asset_info` in a multi_asset with more than one asset. Use"
+                " `context.asset_info_by_key` instead."
+            )
+        if self._asset_info:
+            return self._asset_info
+
+        self._asset_info = AssetInfo(
+            key=self.asset_key,
+            assets_def=self.assets_def,
+            code_version=self.assets_def.code_versions_by_key[self.asset_key],
+            provenance=self.get_asset_provenance(self.asset_key),
+        )
+        return self._asset_info
+
+    @property
+    def asset_info_by_key(self) -> Mapping[AssetKey, AssetInfo]:
+        if self._asset_info_by_key:
+            return self._asset_info_by_key
+
+        self._asset_info_by_key = {}
+        for asset_key in self.selected_asset_keys:
+            self._asset_info_by_key[asset_key] = AssetInfo(
+                key=asset_key,
+                code_version=self.assets_def.code_versions_by_key[asset_key],
+                provenance=self.get_asset_provenance(asset_key),
```

Comment:
`self.get_asset_provenance` involves an instance call, so maybe this should be a (maybe cached) getter method on the `AssetInfo` object rather than pre-fetched here?